File: app/controllers/kernel_controller.rb
class KernelController < ApplicationController
  include ApplicationHelper    
  layout 'standard'
  def ls    
    @page_title = 'Command List (ls)'
    @searching = where(@params['args'])
    @show_golden_egg_labels = true
    # Dinesh G. had the idea of sorting search results by popularity [Jon Aquino 2005-08-13]
    initialize_pagination where(@params['args']), 50, @searching ? 'uses DESC' : 'creation_date DESC'
  end
  def initialize_pagination(where, per_page, order_by)
    @command_pages, @commands = paginate :command, :order_by => order_by, :per_page => per_page, :conditions => where
  end
  def man
    if empty_string_if_nil(@params['args']).empty? then
      redirect_to :action => 'man', :args => 'man'
      return      
    end
    @command_name = @params['args']
    @page_title = @command_name
    @command = Command.find_first ['name=?', @command_name]
    if @command == nil then 
      @page_title = "man #{@command_name}"
      render_action :no_manual_entry 
      return
    end
  end
  def golden_eggs
    @page_title = 'Golden Eggs (ge)'
    @searching = where(@params['args'])
    @show_golden_egg_labels = false
    where = 'golden_egg_date is not null'
    where += ' and (' + where(@params['args']) + ')' if where(@params['args']) != nil
    initialize_pagination(where, @params['all'] == 'true' ? 10000 : 50, @params['all'] == 'true' ? 'name' : @searching ? 'uses DESC' : 'golden_egg_date DESC')
  end
  def most_used_commands
    @page_title = 'The Most-Used Commands'
    @show_golden_egg_labels = true
    initialize_pagination 'last_use_date IS NOT NULL', 50, 'uses DESC'
  end
  def where(args)
    args = empty_string_if_nil(args).strip
    where = nil
    if not args.empty? then
      # Use quote to prevent SQL-injection hacks. Not that it
      # really matters for this application. [Jon Aquino 2005-06-13]
      pattern = "%#{Command.quote(args)[1..-2]}%"
      where = "name like '#{pattern}' or description like '#{pattern}' or url like '#{pattern}'"    
    end
    return where
  end
end

File: app/controllers/KernelController.php
<?php

/**
 * Core functionality
 */
class KernelController extends Controller {

    /** Number of commands per page. */
    const PAGE_SIZE = 50;

    /**
     * Displays the Yubnub homepage.
     *
     * Expected GET parameters:
     *     - page - the page number (optional)
     *     - args - search terms (optional)
     */
    public function action_ls() {
        $page = isset($_GET['page']) ? $_GET['page'] : 1;
        $start = ($page-1) * self::PAGE_SIZE;
        $commandStore = new CommandStore($this->pdoSingleton->getPdo());
        $q = isset($_GET['args']) ? $_GET['args'] : '';
        $extra = 1;
        $commands = $commandStore->findCommands(array(
                'start' => $start,
                'count' => self::PAGE_SIZE + $extra,
                'q' => $q,
                'orderBy' => strlen($q) > 0 ? 'uses DESC' : 'creation_date DESC'));
        $hasNextPage = count($commands) > self::PAGE_SIZE;
        $commands = array_slice($commands, 0, self::PAGE_SIZE - $extra);
        $this->render('ls', array(
            'pageTitle' => 'Command List (ls)',
            'showGoldenEggLabels' => true,
            'searching' => strlen($q) > 0,
            'commands' => $commands,
            'previousPage' => $page > 1 ? $page - 1 : null,
            'nextPage' => $hasNextPage ? $page + 1 : null,
            'previousPageUrl' => $page > 1 ? '?page=' . ($page - 1) . '&args=' . urlencode($q) : null,
            'nextPageUrl' => $hasNextPage ? '?page=' . ($page + 1) . '&args=' . urlencode($q) : null,
        ));
    }

    /**
     * Displays featured commands.
     *
     * Expected GET parameters:
     *     - page - the page number (optional)
     *     - args - search terms (optional)
     */
    public function action_golden_eggs() {
        $page = isset($_GET['page']) ? $_GET['page'] : 1;
        $start = ($page-1) * self::PAGE_SIZE;
        $commandStore = new CommandStore($this->pdoSingleton->getPdo());
        $q = isset($_GET['args']) ? $_GET['args'] : '';
        $extra = 1;
        $commands = $commandStore->findGoldenEggs(array(
                'start' => $start,
                'count' => self::PAGE_SIZE + $extra,
                'q' => $q,
                'orderBy' => strlen($q) > 0 ? 'uses DESC' : 'golden_egg_date DESC'));
        $hasNextPage = count($commands) > self::PAGE_SIZE;
        $commands = array_slice($commands, 0, self::PAGE_SIZE - $extra);
        $this->render('ls', array(
            'pageTitle' => 'Golden Eggs (ge)',
            'showGoldenEggLabels' => false,
            'searching' => strlen($q) > 0,
            'commands' => $commands,
            'previousPage' => $page > 1 ? $page - 1 : null,
            'nextPage' => $hasNextPage ? $page + 1 : null,
            'previousPageUrl' => $page > 1 ? '?page=' . ($page - 1) . '&args=' . urlencode($q) : null,
            'nextPageUrl' => $hasNextPage ? '?page=' . ($page + 1) . '&args=' . urlencode($q) : null,
        ));
    }

    /**
     * Displays a list of the most popular commands.
     *
     * Expected GET parameters:
     *     - page - the page number (optional)
     */
    public function action_most_used_commands() {
        $page = isset($_GET['page']) ? $_GET['page'] : 1;
        $start = ($page-1) * self::PAGE_SIZE;
        $commandStore = new CommandStore($this->pdoSingleton->getPdo());
        $extra = 1;
        $commands = $commandStore->findCommands(array(
                'start' => $start,
                'count' => self::PAGE_SIZE + $extra,
                'orderBy' => 'uses DESC'));
        $hasNextPage = count($commands) > self::PAGE_SIZE;
        $commands = array_slice($commands, 0, self::PAGE_SIZE - $extra);
        $this->render('ls', array(
            'pageTitle' => 'The Most-Used Commands',
            'showGoldenEggLabels' => true,
            'searching' => false,
            'commands' => $commands,
            'previousPage' => $page > 1 ? $page - 1 : null,
            'nextPage' => $hasNextPage ? $page + 1 : null,
            'previousPageUrl' => $page > 1 ? '?page=' . ($page - 1) : null,
            'nextPageUrl' => $hasNextPage ? '?page=' . ($page + 1) : null,
        ));
    }

    /**
     * Displays help for the given command.
     *
     * Expected GET variables:
     *     - args - the name of the command; defaults to 'man'
     */
    public function action_man() {
        $name = isset($_GET['args']) ? $_GET['args'] : null;
        if (strlen($name) == 0) {
            $this->redirectTo('/kernel/man?args=man');
        }
        $commandStore = new CommandStore($this->pdoSingleton->getPdo());
        $command = $commandStore->findCommand($name);
        if (!$command) {
            $this->render('no_manual_entry', array(
                'pageTitle' => 'man ' . $name,
                'name' => $name,
                'lsUrl' => '/parser/parse?command=' . rawurlencode('ls ' . $name)
            ));
        }
        $this->render('man', array(
            'pageTitle' => $name,
            'command' => $command,
        ));
    }

}
